File: day_19/aoc_19_2.py
from typing import Iterable
from collections import defaultdict


def read_input(filename: str) -> tuple[list[str], Iterable[str]]:
    f = open(filename)
    stripes = f.readline().strip().split(', ')

    def _read_designs() -> Iterable[str]:
        while line := f.readline():
            line = line.strip()
            if line:
                yield line
        f.close()

    designs = _read_designs()

    return stripes, designs


def optimize_stripes(stripes: list[str]) -> dict[str, list[str]]:
    stripes_dict = defaultdict(list)

    # Index by the first letter
    for stripe in stripes:
        stripes_dict[stripe[0]].append(stripe)

    # Sort by descending length
    for substripes in stripes_dict.values():
        substripes.sort(key=len, reverse=True)

    return stripes_dict


def solve_designs(stripes: list[str], designs: Iterable[str]) -> int:
    stripes_dict = optimize_stripes(stripes)
    cache = dict()
    return sum(solve_design(stripes_dict, design, cache) for design in designs)


def solve_design(stripes: dict[str, list[str]], design: str, cache: dict[str, int]) -> int:
    if not design:
        return True

    if design in cache:
        return cache[design]

    c = design[0]
    count = 0
    for stripe in stripes[c]:
        # Try with this stripe
        if design.startswith(stripe):
            # Found, keep going
            subdesign = design[len(stripe):]

            # Count all sub-possibilities
            count += solve_design(stripes, subdesign, cache)

    cache[design] = count
    return count


def main(filename: str = 'aoc_19.txt', expected: int = None):
    stripes, designs = read_input(filename)
    result = solve_designs(stripes, designs)

    print(f'{filename=}')
    print(f'{result=}')
    print()

    assert expected is None or result == expected, \
        f'Expected {expected}, got {result}'


if __name__ == '__main__':
    main('aoc_19_example_0.txt', 16)
    main()

File: day_19/aoc_19_2.go
package main

import (
    "bufio"
    "fmt"
    "log"
    "os"
	"strings"
)


func readInput(filename string) ([]string, []string) {
	file, err := os.Open(filename)
    if err != nil {
        log.Fatal(err)
    }
    defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
        log.Fatal(err)
    }

	// Parse the first line
	stripes := strings.Split(lines[0], ", ")

	// The rest of the lines are okay as they are
	return stripes, lines[1:]
}


func optimizeStripes(stripes []string) map[byte][]string {
	stripesDict := make(map[byte][]string)

	// Index by the first letter
	for _, stripe := range stripes {
		letter := stripe[0]

		// Append to the list of stripes starting with that letter
		if prevStripes, exists := stripesDict[letter]; exists {
			stripesDict[letter] = append(prevStripes, stripe)
		} else {
			stripesDict[letter] = []string{stripe}
		}
	}

	return stripesDict
}


func solveDesigns(stripes map[byte][]string, designs []string) int {
	count := 0
	memoization := make(map[string]int)
	for _, design := range designs {
		// Skip empty lines in the input file
		if len(design) > 0 {
			count += solveDesign(stripes, design, memoization)
		}
	}
	return count
}


func solveDesign(stripes map[byte][]string, design string, memoization map[string]int) int {
	// Base case: empty string
	if len(design) == 0 {
		return 1
	}

	// Use cache
	if result, ok := memoization[design]; ok {
		return result
	}

	result := 0

	// Try with all stripes starting with the same letter
	char := design[0]
	if charStripes, ok := stripes[char]; ok {
		for _, stripe := range charStripes {
			if strings.HasPrefix(design, stripe) {
				// This is a potential candidate for solving the design
				nextDesign := design[len(stripe):]
				// Include also the subproblems solution
				result += solveDesign(stripes, nextDesign, memoization)
			}
		}
	}

	// Return the result
	memoization[design] = result
	return result
}


func main() {
	stripes, designs := readInput("aoc_19.txt")
	stripesDict := optimizeStripes(stripes)

	result := solveDesigns(stripesDict, designs)
	fmt.Println(result)
}
